packages/vscode-messenger/src/messenger.ts

```diff
@@ -294,45 +316,58 @@ export class Messenger implements MessengerAPI {
                 } else {
                     return Promise.reject(new Error(`No webview with id ${receiver.webviewId} is registered.`));
                 }
             } else if (receiver.webviewType) {
                 const receiverViews = this.viewTypeRegistry.get(receiver.webviewType);
                 if (receiverViews) {
                     // If there are multiple views, we make a race: the first view to return a result wins
-                    const results = Array.from(receiverViews).map(view => this.sendRequestToWebview(type, receiver, params, view));
+                    const results = Array.from(receiverViews).map(view => this.sendRequestToWebview(type, receiver, params, view, cancelable));
                     return Promise.race(results);
                 } else {
                     return Promise.reject(new Error(`No webview with type ${receiver.webviewType} is registered.`));
                 }
             } else {
                 throw new Error('Unspecified webview receiver: neither webviewId nor webviewType was set.');
             }
-        } else if (receiver.type === 'broadcast') {
-            throw new Error('Only notification messages are allowed for broadcast.');
         }
         throw new Error(`Invalid receiver: ${JSON.stringify(receiver)}`);
     }
 
-    protected async sendRequestToWebview<P, R>(type: RequestType<P, R>, receiver: MessageParticipant, params: P, view: ViewContainer): Promise<R> {
+    protected async sendRequestToWebview<P, R>(type: RequestType<P, R>, receiver: MessageParticipant, params: P, view: ViewContainer, cancelable?: CancellationTokenImpl): Promise<R> {
         // Messages are only delivered if the webview is live (either visible or in the background with `retainContextWhenHidden`).
         if (!view.visible && this.options.ignoreHiddenViews) {
             return Promise.reject(new Error(`Skipped request for hidden view: ${participantToString(receiver)}`));
         }
 
         const msgId = this.createMsgId();
-        const result = new Promise<R>((resolve, reject) => {
-            this.requests.set(msgId, { resolve: resolve as (value: unknown) => void, reject });
-        });
+        const pendingRequest = new Deferred<R>();
+        this.requests.set(msgId, pendingRequest);
+        if (cancelable) {
+            const listener = cancelable.onCancellationRequested((reason) => {
+                // Send cancel message for pending request
+                view.webview.postMessage(createCancelRequestMessage(receiver, { msgId }));
+                pendingRequest.reject(new Error(reason));
+                this.requests.delete(msgId);
+            });
+            pendingRequest.result.finally(() => {
+                // Request finished, remove listener
+                listener.dispose();
+            }).catch((err) =>
+                this.log(`Pending request rejected: ${String(err)}`)
```

Comment:
I think appending `catch` to `finally` has different semantics than the other way around. See also corresponding code in the webview.

Reply:
> I think appending catch to finally has different semantics 

I thought it is basically the same as: 

``` ts
pendingRequest.result.finally(() => {
    // Request finished, remove listener
    listener.dispose();
});
pendingRequest.result.catch((err) =>
    this.log(`Pending request rejected: ${String(err)}`)
);
``` 

Is it not just the execution order of attached callbacks that changes? Which is in this case doesn't matter.

Reply:
I'm not sure. My impression after reading https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally#return_value is that the promise returned by `finally` behaves differently from the original promise.

Reply:
It would be strange, if the kind of how you use the API (directly or chained) would change the behavior. But I will not wonder if it is like that in JS... However, I've tried different order and direct/chained [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally#try_it)  and it behaves the same.